File: 116_Populating_Next_Right_Pointers_in_Each_Node/116_1.c
void connectTwoNode(struct Node *node1, struct Node *node2) {
    if (!node1 || !node2) return;
    node1->next = node2;
    connectTwoNode(node1->left, node1->right);
    connectTwoNode(node2->left, node2->right);
    connectTwoNode(node1->right, node2->left);
}

struct Node* connect(struct Node* root) {
	if (!root) return NULL;
    connectTwoNode(root->left, root->right);
    return root;
}

File: 116_Populating_Next_Right_Pointers_in_Each_Node/116_1.java
class Solution {
    public Node connect(Node root) {
        if (root == null) return root;
        Node pre = root;
        while (pre.left != null) {
            Node cur = pre;
            while (cur != null) {
                cur.left.next = cur.right;
                if (cur.next != null) {
                    cur.right.next = cur.next.left;
                }
                cur = cur.next;
            }
            pre = pre.left;
        }
        return root;
    }
}

